bootstrap/internal/controllers/rke2config_controller.go

```diff
@@ -225,25 +240,329 @@ func (r *Rke2ConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+
+	if !r.RKE2InitLock.Lock(ctx, scope.Cluster, scope.Machine) {
+		scope.Logger.Info("A control plane is already being initialized, requeuing until control plane is ready")
+		return ctrl.Result{RequeueAfter: 1 * time.Minute}, nil
+	}
+
+	defer func() {
+		if reterr != nil {
+			if !r.RKE2InitLock.Unlock(ctx, scope.Cluster) {
+				reterr = kerrors.NewAggregate([]error{reterr, errors.New("failed to unlock the rke2 init lock")})
+			}
+		}
+	}()
+
+	certificates := secret.NewCertificatesForInitialControlPlane()
+	err := certificates.LookupOrGenerate(
+		ctx,
+		r.Client,
+		util.ObjectKey(scope.Cluster),
+		*metav1.NewControllerRef(scope.Config, bootstrapv1.GroupVersion.WithKind("RKE2Config")),
+	)
+	if err != nil {
+		conditions.MarkFalse(scope.Config, bootstrapv1.CertificatesAvailableCondition, bootstrapv1.CertificatesGenerationFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+		return ctrl.Result{}, err
+	}
+	conditions.MarkTrue(scope.Config, bootstrapv1.CertificatesAvailableCondition)
+
+	token, err := r.generateAndStoreToken(ctx, scope)
+	if err != nil {
+		scope.Logger.Error(err, "unable to generate and store an RKE2 server token")
+		return ctrl.Result{}, err
+	}
+	scope.Logger.Info("RKE2 server token generated and stored in Secret!")
+
+	configStruct, files, err := rke2.GenerateInitControlPlaneConfig(
+		rke2.RKE2ServerConfigOpts{
+			ControlPlaneEndpoint: scope.Cluster.Spec.ControlPlaneEndpoint.Host,
+			Token:                token,
+			ServerURL:            "https://" + scope.Cluster.Spec.ControlPlaneEndpoint.Host + ":9345",
+			ServerConfig:         scope.ControlPlane.Spec.ServerConfig,
+			AgentConfig:          scope.Config.Spec.AgentConfig,
+			Ctx:                  ctx,
+			Client:               r.Client,
+		})
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	b, err := kubeyaml.Marshal(configStruct)
+	if err != nil {
+
+		return ctrl.Result{}, err
+	}
+	scope.Logger.Info("Server config marshalled successfully")
+
+	initConfigFile := bootstrapv1.File{
+		Path:        rke2.DefaultRKE2ConfigLocation,
+		Content:     string(b),
+		Owner:       "root:root",
+		Permissions: "0640",
+	}
+
+	// TODO: Implement adding additional files through API
+
+	cpinput := &cloudinit.ControlPlaneInput{
+		BaseUserData: cloudinit.BaseUserData{
+			AirGapped:        scope.Config.Spec.AgentConfig.AirGapped,
+			PreRKE2Commands:  scope.Config.Spec.PreRKE2Commands,
+			PostRKE2Commands: scope.Config.Spec.PostRKE2Commands,
+			ConfigFile:       initConfigFile,
+			RKE2Version:      scope.Config.Spec.AgentConfig.Version,
+			WriteFiles:       files,
+		},
+		Certificates: certificates,
+	}
+
+	cloudInitData, err := cloudinit.NewInitControlPlane(cpinput)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.storeBootstrapData(ctx, scope, cloudInitData); err != nil {
+		return ctrl.Result{}, err
+	}
 
-func Unlock(ctx context.Context, cluster *clusterv1.Cluster) bool {
-	return true
+	return ctrl.Result{}, nil
```

Comment:
This `handleClusterNotInitialized` function is doing quite a lot within it. I wonder if there is a cleaner way to represent this

Reply:
Especially considering we will want to add ignition support in the future.

Reply:
But we can refactor this at a later date.....its just a nit.

Reply:
Well, totally open to that, especially if we do implement Ignition. However, looking at the code, it does look like the function has clear sections:
- Preliminary checks
- defer func
- handle certificates
- handle token
- handle config.yaml creation
- handle cloud-init

I could outsource the cloud-init part to another func, but the other sections look like they are already outsourced. Do you agree ?

Reply:
Let's keep it so for now, and refactor the last section when we implement the ignition part.